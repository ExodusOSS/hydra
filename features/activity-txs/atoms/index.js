export { default as activityTxsAtomDefinition } from './activity-txs'
