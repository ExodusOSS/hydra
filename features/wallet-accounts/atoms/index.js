import createEnabledWalletAccountsAtom from './enabled-wallet-accounts.js'
import createWalletAccountsAtom from './wallet-accounts.js'
import createActiveWalletAccountAtom from './active-wallet-account.js'
import createMultipleWalletAccountsEnabledAtom from './multiple-wallet-accounts-enabled.js'
import createHardwareWalletPublicKeysAtom from './hardware-wallet-public-keys.js'
import createWalletAccountsInternalAtom from './wallet-accounts-internal.js'

export const walletAccountsAtomDefinition = {
  id: 'walletAccountsAtom',
  type: 'atom',
  factory: createWalletAccountsAtom,
  dependencies: ['walletAccountsInternalAtom'],
  public: true,
}

export const walletAccountsInternalAtomDefinition = {
  id: 'walletAccountsInternalAtom',
  type: 'atom',
  factory: createWalletAccountsInternalAtom,
  dependencies: ['storage', 'config'],
}

export const enabledWalletAccountsAtomDefinition = {
  id: 'enabledWalletAccountsAtom',
  type: 'atom',
  factory: createEnabledWalletAccountsAtom,
  dependencies: ['walletAccountsAtom'],
  public: true,
}

export const activeWalletAccountAtomDefinition = {
  id: 'activeWalletAccountAtom',
  type: 'atom',
  factory: createActiveWalletAccountAtom,
  dependencies: ['storage'],
  public: true,
}

export const multipleWalletAccountsEnabledAtomDefinition = {
  id: 'multipleWalletAccountsEnabledAtom',
  type: 'atom',
  factory: createMultipleWalletAccountsEnabledAtom,
  dependencies: ['fusion', 'logger'],
  public: true,
}

export const hardwareWalletPublicKeysAtomDefinition = {
  id: 'hardwareWalletPublicKeysAtom',
  type: 'atom',
  factory: createHardwareWalletPublicKeysAtom,
  dependencies: ['storage'],
  public: true,
}
