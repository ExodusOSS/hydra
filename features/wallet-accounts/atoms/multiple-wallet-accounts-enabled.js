import { createFusionAtom } from '@exodus/fusion-atoms'

const createMultipleWalletAccountsEnabledAtom = ({ fusion, logger }) =>
  createFusionAtom({
    fusion,
    logger,
    path: 'multipleWalletAccountsEnabled',
    defaultValue: undefined,
  })

export default createMultipleWalletAccountsEnabledAtom
