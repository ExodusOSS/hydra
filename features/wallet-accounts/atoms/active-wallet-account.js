import { createStorageAtomFactory } from '@exodus/atoms'
import { WalletAccount } from '@exodus/models'

const createActiveWalletAccountAtom = ({ storage }) => {
  const atomFactory = createStorageAtomFactory({ storage })

  return atomFactory({
    key: 'activeWalletAccount',
    defaultValue: WalletAccount.DEFAULT_NAME,
    isSoleWriter: true,
  })
}

export default createActiveWalletAccountAtom
