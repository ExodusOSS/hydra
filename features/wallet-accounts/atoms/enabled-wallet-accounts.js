import { compute } from '@exodus/atoms'

export default function createEnabledWalletAccountsAtom({ walletAccountsAtom }) {
  const selector = (values) =>
    Object.fromEntries(Object.entries(values).filter(([, account]) => account.enabled))

  return compute({ atom: walletAccountsAtom, selector })
}
