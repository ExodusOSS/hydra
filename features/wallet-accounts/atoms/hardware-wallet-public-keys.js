import { createStorageAtomFactory } from '@exodus/atoms'
import { createEmptyAccounts } from '../module/utils.js'

export default function createHardwareWalletPublicKeysAtom({ storage }) {
  const atomFactory = createStorageAtomFactory({ storage })

  return atomFactory({
    key: 'hardwareWalletPublicKeys',
    defaultValue: createEmptyAccounts(),
    isSoleWriter: true,
  })
}
