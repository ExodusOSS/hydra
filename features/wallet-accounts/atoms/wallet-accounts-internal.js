import { createStorageAtomFactory, dedupe, withSerialization } from '@exodus/atoms'
import { mapValues } from '@exodus/basic-utils'
import lodash from 'lodash'
import { WalletAccount } from '@exodus/models'

const { isEmpty } = lodash

export default function createWalletAccountsInternalAtom({
  storage,
  config = Object.create(null),
}) {
  const atomFactory = createStorageAtomFactory({ storage })

  const initialWalletAccounts = {
    [WalletAccount.DEFAULT_NAME]: {
      ...WalletAccount.DEFAULT,
      label: config.defaultLabel,
      color: config.defaultColor,
    },
  }

  const walletAccountsAtom = atomFactory({
    key: 'walletAccounts',
    defaultValue: initialWalletAccounts,
    isSoleWriter: true,
  })

  const deserialize = (input) => {
    if (!input || isEmpty(input)) {
      console.warn(
        'WalletAccountAtom: Invalid atom value. Using default, but this should not happen.'
      )
      return initialWalletAccounts
    }

    const walletAccounts = mapValues(input, (data) => new WalletAccount(data))

    // Ensure that normal wallet accounts are sorted before hardware wallet accounts
    const sortedWalletAccounts = Object.entries(walletAccounts).sort(([, a], [, b]) => {
      return a.isHardware - b.isHardware
    })

    return Object.fromEntries(sortedWalletAccounts)
  }

  const serialize = (input) => {
    if (!input || isEmpty(input)) {
      // support `atom.set(undefined)` for clearing storage
      return
    }

    return mapValues(input, (data) => data.toJSON())
  }

  return dedupe(withSerialization({ atom: walletAccountsAtom, serialize, deserialize }))
}
