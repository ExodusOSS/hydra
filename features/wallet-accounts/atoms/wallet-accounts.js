import { filter } from '@exodus/atoms'
import { WalletAccount } from '@exodus/models'

export default function createWalletAccountsAtom({ walletAccountsInternalAtom }) {
  return filter(walletAccountsInternalAtom, (value) => !!value[WalletAccount.DEFAULT_NAME]?.seedId)
}
