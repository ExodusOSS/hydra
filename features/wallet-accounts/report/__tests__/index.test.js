import { createInMemoryAtom } from '@exodus/atoms'
import { WalletAccount } from '@exodus/models'

import walletAccountsReportDefinition from '../index.js'

describe('walletAccountsReport', () => {
  let walletAccountsAtom
  let activeWalletAccountAtom
  let multipleWalletAccountsEnabledAtom

  const walletAccountsData = {
    exodus_0: new WalletAccount({
      color: '#ff3974',
      enabled: true,
      icon: 'exodus',
      index: 0,
      label: 'Exodus',
      source: 'exodus',
    }),
    exodus_1: new WalletAccount({
      color: '#30d968',
      enabled: true,
      icon: 'trezor',
      index: 1,
      label: 'asdf',
      source: 'exodus',
    }),
  }

  beforeEach(() => {
    walletAccountsAtom = createInMemoryAtom({ defaultValue: walletAccountsData })
    activeWalletAccountAtom = createInMemoryAtom({ defaultValue: 'exodus_0' })
    multipleWalletAccountsEnabledAtom = createInMemoryAtom({ defaultValue: undefined })
  })

  it('should provide the correct namespace', async () => {
    const report = walletAccountsReportDefinition.factory({
      walletAccountsAtom,
      activeWalletAccountAtom,
      multipleWalletAccountsEnabledAtom,
    })

    expect(report.namespace).toEqual('walletAccounts')
  })

  it('should report wallet accounts', async () => {
    const report = walletAccountsReportDefinition.factory({
      walletAccountsAtom,
      activeWalletAccountAtom,
      multipleWalletAccountsEnabledAtom,
    })
    const result = await report.export()

    const data = Object.fromEntries(
      Object.entries(walletAccountsData).map(([key, walletAccount]) => [
        key,
        new WalletAccount({ ...walletAccount, label: '<Redacted>' }),
      ])
    )

    expect(result).toEqual({
      data,
      configuredActiveWalletAccount: 'exodus_0',
      multipleWalletAccountsEnabled: undefined,
    })
  })
})
