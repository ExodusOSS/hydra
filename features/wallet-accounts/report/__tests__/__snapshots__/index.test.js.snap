exports[`walletAccountsReport should report wallet accounts 1`] = `
{
  "configuredActiveWalletAccount": "exodus_0",
  "data": {
    "exodus_0": {
      "enabled": true,
      "index": 0,
      "isMultisig": false,
      "source": "exodus",
    },
    "exodus_1": {
      "enabled": true,
      "index": 1,
      "isMultisig": false,
      "source": "exodus",
    },
  },
  "multipleWalletAccountsEnabled": undefined,
}
`;
